### PR TITLE
Fix evaluation of dominant response on MakeSCRestAPICall

### LINKED ISSUE #534: response body not returned as error when dominant sharder response is not 200
response body not returned as error when dominant sharder response is not 200.

https://github.com/0chain/gosdk/blob/70a96ae85bff57245604f5ff6cd7e262519074f4/zboxcore/zboxutil/http.go#L570

the content of error was being set from last received response from sharder. for example, below shows the added error was from good response. the next example is able to show error correctly

**with good response as error**

> allocation_fetch_error: Error fetching the allocation.{"id":"41ab2b1715356a8e5df41b354ad47ed59e42827721cd822fe9972d4d37966d3f","tx":"41ab2b1715356a8e5df41b354ad47ed59e42827721cd822fe9972d4d37966d3f","data_shards":1,"parity_shards":1,"size":10485760,"expiration_date":1660232463,"owner_id":"69c262886ce856ca8397d6fceb8b6388a8c8cda7c9424eafa96ab58a930e19f5","owner_public_key":"0468ef36a1e7879182b8f4fcaf7c09b7316701c5ffe504b44f554a622075780b2ad344cba881e42fe84d0a0f133fa39fedb714462b333f04c1e97b9705e8b00c","stats":{"used_size":0,"num_of_writes":0,"num_of_reads":0,"total_challenges":0,"num_open_challenges":0,"num_success_challenges":0,"num_failed_challenges":0,"latest_closed_challenge":""},"diverse_blobbers":false,"preferred_blobbers":null,"blobber_details":[{"blobber_id":"c50ebe28f8a8aad4e9205030533765887f4449b9cf250ad0315c9d4afa9b4ef7","allocation_id":"41ab2b1715356a8e5df41b354ad47ed59e42827721cd822fe9972d4d37966d3f","size":16038821300,"allocation_root":"","write_marker":null,"stats":null,"terms":{"read_price":100000000,"write_price":1000000000,"min_lock_demand":0.1,"max_offer_duration":2678400000000000},"min_lock_demand":16276,"spent":0,"penalty":0,"read_reward":0,"returned":0,"challenge_reward":0,"final_reward":0,"challenge_pool_integral_value":0,"blobber_allocs_partition_loc":null},{"blobber_id":"eae9a0ea2fd06645f8cdb3175a23999054f01ae1a7edd06847acd0548cbeee78","allocation_id":"41ab2b1715356a8e5df41b354ad47ed59e42827721cd822fe9972d4d37966d3f","size":20463812020,"allocation_root":"","write_marker":null,"stats":null,"terms":{"read_price":100000000,"write_price":1000000000,"min_lock_demand":0.1,"max_offer_duration":2678400000000000},"min_lock_demand":16276,"spent":0,"penalty":0,"read_reward":0,"returned":0,"challenge_reward":0,"final_reward":0,"challenge_pool_integral_value":0,"blobber_allocs_partition_loc":null}],"is_immutable":false,"third_party_extendable":false,"file_options":0,"write_pool":5000000000,"read_price_range":{"min":0,"max":9223372036854775807},"write_price_range":{"min":0,"max":9223372036854775807},"challenge_completion_time":0,"start_time":1660146063,"time_unit":2592000000000000,"curators":[],"name":"","blobbers":[{"id":"c50ebe28f8a8aad4e9205030533765887f4449b9cf250ad0315c9d4afa9b4ef7","url":"https://hightps-2.devnet-0chain.net/blobber02","geolocation":{"latitude":0,"longitude":0},"terms":{"read_price":100000000,"write_price":1000000000,"min_lock_demand":0.1,"max_offer_duration":2678400000000000},"capacity":337589997568,"allocated":16038821300,"last_health_check":1660146151,"saved_data":0,"data_read_last_reward_round":0,"last_reward_data_read_round":0,"stake_pool_settings":{"delegate_wallet":"9c693cb14f29917968d6e8c909ebbea3425b4c1bc64b6732cadc2a1869f49be9","min_stake":10000000000,"max_stake":1000000000000,"num_delegates":50,"service_charge":0.3},"reward_partition":{"index":0,"start_round":0,"timestamp":0},"info":{"name":"","website_url":"","logo_url":"","description":""}},{"id":"eae9a0ea2fd06645f8cdb3175a23999054f01ae1a7edd06847acd0548cbeee78","url":"https://hightps-2.devnet-0chain.net/blobber04","geolocation":{"latitude":0,"longitude":0},"terms":{"read_price":100000000,"write_price":1000000000,"min_lock_demand":0.1,"max_offer_duration":2678400000000000},"capacity":337515417600,"allocated":20463812020,"last_health_check":1660146177,"saved_data":0,"data_read_last_reward_round":0,"last_reward_data_read_round":0,"stake_pool_settings":{"delegate_wallet":"9c693cb14f29917968d6e8c909ebbea3425b4c1bc64b6732cadc2a1869f49be9","min_stake":10000000000,"max_stake":1000000000000,"num_delegates":50,"service_charge":0.3},"reward_partition":{"index":0,"start_round":0,"timestamp":0},"info":{"name":"","website_url":"","logo_url":"","description":""}}]}

**with actual error**

> allocation_fetch_error: Error fetching the allocation.internal_error: can't get allocation: error retrieving allocation: 47e4250e012b40d914aea81be8b5bbfcd0befb98463a5b2b9ff761bb7ea762af, error: failed to connect to `host=helm-sharder-postgres-02 user=zchain_user database=events_db`: server error (FATAL: remaining connection slots are reserved for non-replication superuser connections (SQLSTATE 53300))

### PR DESCRIPTION
**Pull request by avanaur**
### Changes

Fix evaluation of dominant response on MakeSCRestAPICall. 

The fix ensures that the appropriate response body is returned when there is a mix of different responses from sharders. 
The change ensures the response of dominant status is returned and in case of tie-breakers will prioritize 200 response.

Current, the first to reach highest count was being set as candidate response and there is no tie-breaker check. On tie-breaker situation, the dominant status selected can be different to candidate response resulting to bad parsing on error or good data. (see error samples on https://github.com/0chain/gosdk/issues/534)

### Fixes
Resolves https://github.com/0chain/gosdk/issues/534

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/gosdk/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- blobber:
- 0chain:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: zboxcore/zboxutil/http.go [hunk 1]
```diff
@@ -542,9 +543,18 @@ func MakeSCRestAPICall(scAddress string, relativePath string, params map[string]
 				responses[response.StatusCode]++
 				if responses[response.StatusCode] > maxCount {
 					maxCount = responses[response.StatusCode]
+				}
+
+				if responses[response.StatusCode] > maxCount {
```
Comment (dabasov) [inline zboxcore/zboxutil/http.go]: why is it duplicated?

**Review by dabasov — DISMISSED** [review]

File: zboxcore/zboxutil/http.go [hunk 2]
```diff
@@ -519,6 +519,7 @@ func MakeSCRestAPICall(scAddress string, relativePath string, params map[string]
 	entityResult := make(map[string][]byte)
 	var retObj []byte
 	maxCount := 0
+	dominant := 200
```
Comment (stewartie4) [inline zboxcore/zboxutil/http.go:548]: could this be more descriptive : dominantResponseCode?

File: zboxcore/zboxutil/http.go [hunk 3]
```diff
@@ -542,9 +543,18 @@ func MakeSCRestAPICall(scAddress string, relativePath string, params map[string]
 				responses[response.StatusCode]++
 				if responses[response.StatusCode] > maxCount {
 					maxCount = responses[response.StatusCode]
+				}
+
+				if responses[response.StatusCode] > maxCount {
+					maxCount = responses[response.StatusCode]
+				}
+
+				if isCurrentDominantStatus(response.StatusCode, responses, maxCount) {
```
Comment (stewartie4) [inline zboxcore/zboxutil/http.go:574]: Presumably we will also want to do this on every other http call that requires consensus, including calls to miners?

**Review by stewartie4 — APPROVED** [review]
Happy to approve pending duplicate code deletion

**Review by peterlimg — APPROVED** [review]
lgtm

File: zboxcore/zboxutil/http.go (hunk 1 above)
Comment (avanaur) [reply zboxcore/zboxutil/http.go]: Just a merge issue. Removed the duplicate

File: zboxcore/zboxutil/http.go (hunk 3 above)
Comment (avanaur) [reply zboxcore/zboxutil/http.go:574]: @stewartie4 I did not look into other parts as I am not sure what are the "actual rules". 

At one point, I actually thought that it should not always use dominant response. For example, if threshold consensus is 25% and if 1 out of 4 of response is 200, then maybe that is what should be used as response even if 3 out of 4 responses is 404. FYI @dabasov @peterlimg 

The PR is just to fix what "seems" intended by function by the current implementation.

**Review by stewartie4 — APPROVED** [review]
LGTM

File: zboxcore/zboxutil/http.go (hunk 3 above)
Comment (lpoli) [reply zboxcore/zboxutil/http.go:574]: No we should not do that.
If 3 out of 4 responses are 404 then we should return 404 error.
cc @peterlimg

**Pull request merged**